src/utils/futures.rs

```diff
@@ -280,6 +280,39 @@ pub async fn delay(duration: Duration) {
         .ok();
 }
 
+/// Time a future execution and expose using metrics.
+///
+/// The future is expected to resolve to a Result and the status is logged as well.
+pub async fn timed_task<F, T, E>(task_name: &str, future: F) -> F::Output
+where
+    F: Future<Output = Result<T, E>>,
+{
+    let start_time = Instant::now();
+    let ret = future.await;
+    metric!(
+        timer("futures.done") = start_time.elapsed(),
+        "task_name" => task_name,
+        "status" => if ret.is_ok() { "ok" } else {"err"},
+    );
+    ret
+}
+
+pub trait TimedTaskExt<T, E>: Future<Output = Result<T, E>> + Sized {
```

Comment:
I understand this might be controversial.  I think it reads nicer when using it, but don't feel overly strong.

Reply:
This sure is controversial. I do think that chaining combinators is a lot nicer to read than wrapping in functions.
The problem right now is that it is both super ugly to implement, and less performant due to boxing. I also predict that stable async-fn-in-traits will probably take >1 year :-(

Reply:
Is boxing really that terrible?  We do it all over the place now with the futures01/03 migration.  Doesn't seem to cause that much issues.

Reply:
In the long run, it would be nicer to get rid of Boxes around futures. We mostly add them where they help with keeping types small or when you have to join types from two paths. Also adding a comment on this specific signature.